drivers/gpio/gpio_nrfx.c

```diff
@@ -0,0 +1,455 @@
+{
+	NRF_GPIO_Type *reg = get_port_cfg(port)->port;
+	struct gpio_nrfx_data *data = get_port_data(port);
+
+	if (access_op == GPIO_ACCESS_BY_PORT) {
+		nrf_gpio_port_out_write(reg, value ^ data->polarity);
+	} else {
+		if ((value > 0) ^ ((BIT(pin) & data->polarity) != 0)) {
+			nrf_gpio_port_out_set(reg, BIT(pin));
+		} else {
+			nrf_gpio_port_out_clear(reg, BIT(pin));
+		}
+	}
+
+	return 0;
+}
+
+static int gpio_nrfx_read(struct device *port, int access_op,
+			  u32_t pin, u32_t *value)
+{
+	NRF_GPIO_Type *reg = get_port_cfg(port)->port;
+	struct gpio_nrfx_data *data = get_port_data(port);
+
+	if (access_op == GPIO_ACCESS_BY_PORT) {
+		*value = nrf_gpio_port_in_read(reg);
+		*value ^= data->polarity;
+	} else {
+		*value = (nrf_gpio_port_in_read(reg) & BIT(pin)) >> pin;
+		*value = (BIT(pin) & data->polarity) ? !(*value) : *value;
+	}
+
+	return 0;
+}
+
+static int gpio_nrfx_manage_callback(struct device *port,
+				     struct gpio_callback *callback,
+				     bool set)
+{
+	_gpio_manage_callback(&get_port_data(port)->callbacks, callback, set);
+	return 0;
+}
+
+static int gpio_nrfx_enable_callback(struct device *port,
+				     int access_op,
+				     u32_t pin)
+{
+	struct gpio_nrfx_data *data = get_port_data(port);
+	u8_t pins_affected = 1;
+	int res = 0;
+
+	if (access_op == GPIO_ACCESS_BY_PORT) {
+		pins_affected = popcount(pin);
+	} else {
+		pin = BIT(pin);
+	}
+
+	for (size_t i = 0; i < pins_affected; i++) {
+		u32_t cur_pin = find_lsb_set(pin) - 1;
+
+		pin &= ~BIT(cur_pin);
+
+		if (!(data->pin_int_en & BIT(cur_pin))) {
+			return -EINVAL;
+		}
+
+		data->int_en |= BIT(cur_pin);
+
+		res = gpiote_pin_int_cfg(port, cur_pin);
+		if (res != 0) {
+			return res;
+		}
+	}
+
+	return res;
+}
+
+static int gpio_nrfx_disable_callback(struct device *port,
+				      int access_op,
+				      u32_t pin)
+{
+	struct gpio_nrfx_data *data = get_port_data(port);
+	u8_t pins_affected = 1;
+	int res = 0;
+
+	if (access_op == GPIO_ACCESS_BY_PORT) {
+		pins_affected = popcount(pin);
+	} else {
+		pin = BIT(pin);
+	}
+
+	for (size_t i = 0; i < pins_affected; i++) {
+		u32_t cur_pin = find_lsb_set(pin) - 1;
+
+		pin &= ~BIT(cur_pin);
+
+		data->int_en &= ~BIT(cur_pin);
+
+		res = gpiote_pin_int_cfg(port, cur_pin);
+		if (res != 0) {
+			return res;
+		}
+	}
+
+	return res;
+}
+
+static const struct gpio_driver_api gpio_nrfx_drv_api_funcs = {
+	.config = gpio_nrfx_config,
+	.write = gpio_nrfx_write,
+	.read = gpio_nrfx_read,
+	.manage_callback = gpio_nrfx_manage_callback,
+	.enable_callback = gpio_nrfx_enable_callback,
+	.disable_callback = gpio_nrfx_disable_callback
+};
+
+static u32_t get_trigger_pins(struct device *port)
+{
+	if (port == NULL) {
+		return 0;
+	}
+	struct gpio_nrfx_data *data = get_port_data(port);
+
+	u32_t port_in = nrf_gpio_port_in_read(get_port_cfg(port)->port);
+	u32_t xor = (port_in ^ data->active_level ^ data->polarity);
+	u32_t out = ~xor & data->int_en & data->pin_int_en
+		    & ~data->trig_edge & ~data->double_edge;
```

Comment:
I'm afraid this will be incomprehensible for the majority of sane people. Add at least a word of comment, please.